src/core/colorspace_utils.js

```diff
@@ -0,0 +1,296 @@
+      cs,
+      xref,
+      resources,
+      pdfFunctionFactory
+    );
+
+    // Attempt to cache the parsed ColorSpace, by name and/or reference.
+    this._cache(cs, xref, localColorSpaceCache, parsedColorSpace);
+
+    return parsedColorSpace;
+  }
+
+  /**
+   * @private
+   */
+  static _parse(cs, xref, resources = null, pdfFunctionFactory) {
+    cs = xref.fetchIfRef(cs);
+    if (cs instanceof Name) {
+      switch (cs.name) {
+        case "G":
+        case "DeviceGray":
+          return this.singletons.gray;
+        case "RGB":
+        case "DeviceRGB":
+          return this.singletons.rgb;
+        case "DeviceRGBA":
+          return this.singletons.rgba;
+        case "CMYK":
+        case "DeviceCMYK":
+          return this.singletons.cmyk;
+        case "Pattern":
+          return new PatternCS(/* baseCS = */ null);
+        default:
+          if (resources instanceof Dict) {
+            const colorSpaces = resources.get("ColorSpace");
+            if (colorSpaces instanceof Dict) {
+              const resourcesCS = colorSpaces.get(cs.name);
+              if (resourcesCS) {
+                if (resourcesCS instanceof Name) {
+                  return this._parse(
+                    resourcesCS,
+                    xref,
+                    resources,
+                    pdfFunctionFactory
+                  );
+                }
+                cs = resourcesCS;
+                break;
+              }
+            }
+          }
+          // Fallback to the default gray color space.
+          warn(`Unrecognized ColorSpace: ${cs.name}`);
+          return this.singletons.gray;
+      }
+    }
+    if (Array.isArray(cs)) {
+      const mode = xref.fetchIfRef(cs[0]).name;
+      let params, numComps, baseCS, whitePoint, blackPoint, gamma;
+
+      switch (mode) {
+        case "G":
+        case "DeviceGray":
+          return this.singletons.gray;
+        case "RGB":
+        case "DeviceRGB":
+          return this.singletons.rgb;
+        case "CMYK":
+        case "DeviceCMYK":
+          return this.singletons.cmyk;
+        case "CalGray":
+          params = xref.fetchIfRef(cs[1]);
+          whitePoint = params.getArray("WhitePoint");
+          blackPoint = params.getArray("BlackPoint");
+          gamma = params.get("Gamma");
+          return new CalGrayCS(whitePoint, blackPoint, gamma);
+        case "CalRGB":
+          params = xref.fetchIfRef(cs[1]);
+          whitePoint = params.getArray("WhitePoint");
+          blackPoint = params.getArray("BlackPoint");
+          gamma = params.getArray("Gamma");
+          const matrix = params.getArray("Matrix");
+          return new CalRGBCS(whitePoint, blackPoint, gamma, matrix);
+        case "ICCBased":
+          const stream = xref.fetchIfRef(cs[1]);
+          const dict = stream.dict;
+          numComps = dict.get("N");
+
+          if (IccColorSpace.isUsable) {
+            return new IccColorSpace(stream.getBytes(), numComps);
```

Comment:
We should ensure that any Errors thrown from `IccColorSpace` doesn't completely break rendering, since falling back to the handling below means that we'll still render *something* rather than e.g. an entirely broken/invisible image.
```suggestion
            try {
              return new IccColorSpace(stream.getBytes(), numComps);
            } catch (ex) {
              warn(`ICCBased color space: "{ex}".`);
            }
```